### ISSUE TITLE
Implement restrictions

### ISSUE THREAD

**Issue by oxalorg** [oxalorg/4ever-clojure#16]
- [ ] Configure such that an error is thrown for restricted functions, probably by using sci `deny`
- [x] Show a warning for restricted functions - #14 solves this

**Comment by pradeepbishnoi** [comment]
To `:deny` use below example as reference
```(eval-string "(first '(1 2 3 4))" {:deny '[inc dec first]})```

**Comment by pradeepbishnoi** [comment]
@borkdude Any suggestion how to convert this into a quote for 
Tried few things however not able to convert it.
```(first `(~(mapv symbol ["inc" "dec" "last"])))```

above expression give a vector with symbol. however want to convert this into a quoted collection for input to :deny.

<img width="498" alt="image" src="https://user-images.githubusercontent.com/1634146/125185118-64b06c00-e240-11eb-9714-b61f11d132f0.png">

**Comment by borkdude** [comment]
Just `(mapv symbol ...)` should do the trick? Why are you using the backtick?

**Comment by pradeepbishnoi** [comment]
@borkdude Yes, it does. I over thought on this one and didn't realise that :deny is just needing a vector. Using the backpack I was trying to create a "quoted" statement to pass 🤦

**Comment by Invertisment** [comment]
You're overcomplicating things.
The thing that the user enters is an EDN string. 

I wrote this function by trying to solve this problem and as I currently don't yet have a proper branch to merge it from so I post it here:

```
(defn find-restricted-symbols [restricted-symbol-strings user-input]
  (let [restricted-symbols (set (map symbol restricted-symbol-strings))]
    (->> user-input
         clojure.edn/read-string
         vector
         (clojure.walk/prewalk (fn [form]
                                 (cond (map? form) (concat (keys form) (vals form))
                                       (set? form) (seq form)
                                       :else form)))
         flatten
         set
         (clojure.set/intersection restricted-symbols)
         seq)))

#_(find-restricted-symbols ["flatten" "into"] "(flatten (into [] [[1]]))")
#_(find-restricted-symbols ["flatten" "into"] "#{:my-set-item flatten}")
#_(find-restricted-symbols ["flatten" "into"] "{}")
#_(find-restricted-symbols ["flatten" "into"] "{flatten []}")
#_(find-restricted-symbols ["flatten" "into"] "{flatten {:my-map-key into} }")
#_(find-restricted-symbols ["flatten" "into"] "{flatten {:my-map-key #{[into]}} }")
#_(find-restricted-symbols ["flatten" "into"] "flatten")
#_(find-restricted-symbols ["max" "max-key"] "[max \"max-key\"]")

```

**Comment by borkdude** [comment]
What about just using `str/includes?` on the input string? ;)

**Comment by Invertisment** [comment]
> What about just using str/includes? on the input string? ;)

Yes it would work but my function distinguishes between `"max-key"` and `max-key`

**Comment by borkdude** [comment]
No, that would give false positives on locals with that name. The `:deny` option in SCI was built to do this correctly, only on var names.

**Comment by Invertisment** [comment]
My function would fail on `(fn max-key []...)`, true. And NS resolution wouldn't help.

Would SCI matching recover from this correctly?

**Comment by borkdude** [comment]
yes

**Comment by borkdude** [comment]
```
user=> (sci/eval-string "((fn max-key [] 1))" {:deny '[max-key]})
1
user=> (sci/eval-string "(max-key (fn max-key [] 1))" {:deny '[max-key]})
Execution error (ExceptionInfo) at sci.impl.utils/throw-error-with-location (utils.cljc:41).
max-key is not allowed!
```

**Comment by Invertisment** [comment]
Should it still evaluate and show the results together with restriction error?

**Comment by borkdude** [comment]
@Invertisment Maybe we can make restricting a setting like you did for the editor settings. In case an error happens with restrictions, we want to offer the ability to turn that off.

**Comment by Invertisment** [comment]
Do you mean that you would expect to accept `range` as solution to "implement range" if this kind of option would be false?
I was thinking about simply showing additional message.

Also if restriction check wouldn't be part of evaluator then we could do this check elsewhere.
But yes. If it would be part of evaluator then we would either need to evaluate it twice or only show either an error or result but nothing else.

**Comment by borkdude** [comment]
Evaluating twice also seems fine, in case the restriction is hit.
